### PR TITLE
cross-dataset: Layer-wise Learning Rate Decay (LLRD) across all 4 datasets

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

**Layer-wise Learning Rate Decay (LLRD)** applies a different learning rate to each transformer layer, with earlier (input-side) layers receiving a smaller LR and later (output-side) layers receiving a larger LR. This is well-established in NLP fine-tuning (BERT, ViT) and the intuition transfers to CFD surrogates: early layers encode fundamental geometric structure that should change slowly, while later layers perform task-specific pressure prediction that benefits from faster adaptation.

Standard single-LR training may cause early layers to overfit or oscillate while later layers are still converging, or vice versa. LLRD provides a natural regularization gradient across depth.

**Proposed decay factor**: `lr[layer_i] = base_lr * decay^(n_layers - 1 - i)` where `decay=0.8` means each layer gets 80% of the next layer's LR.

## Current Baselines (lower is better)

| Dataset | Metric | Value | PR |
|---|---|---|---|
| tandemfoil | val_primary/surface_pressure_mae | **26.06** | #2887 |
| airfrans | val_primary/surface_mse | **0.000598** | #2906 |
| drivaerml | val_primary/surface_rel_l2_pct | **3.997%** | #2898 |
| tandemfoil_paper | val_primary/field_mse | no baseline yet | — |

## Instructions to Student

**This requires a CODE CHANGE** to `train.py` or the optimizer setup in `core/`. The optimizer must receive per-layer parameter groups with different learning rates.

Replace the single optimizer initialization with grouped parameter configuration:

```python
def get_llrd_param_groups(model, base_lr, decay=0.8):
    """
    Build optimizer param groups with layer-wise LR decay.
    Returns list of param group dicts suitable for AdamW/Lion.
    """
    # Identify transformer layers in order (input to output)
    # Adjust these names to match the actual model attribute names
    layer_params = []
    
    # Example: model has model.layers = nn.ModuleList([...])
    n_layers = len(model.layers)
    for i, layer in enumerate(model.layers):
        lr_i = base_lr * (decay ** (n_layers - 1 - i))
        layer_params.append({
            'params': list(layer.parameters()),
            'lr': lr_i,
            'name': f'layer_{i}'
        })
    
    # Embedding / input layers get the smallest LR (decay^n_layers * base_lr)
    embed_lr = base_lr * (decay ** n_layers)
    other_params = [p for n, p in model.named_parameters()
                    if not any(f'layers.{i}.' in n for i in range(n_layers))]
    layer_params.insert(0, {'params': other_params, 'lr': embed_lr, 'name': 'embed'})
    
    return layer_params

# Then:
param_groups = get_llrd_param_groups(model, base_lr=args.lr, decay=0.8)
optimizer = torch.optim.AdamW(param_groups, weight_decay=args.weight_decay)
# or for Lion:
# optimizer = Lion(param_groups, weight_decay=args.weight_decay)
```

Test two decay values in parallel: **decay=0.8** and **decay=0.9** on TandemFoil first. If one clearly dominates, use that for the remaining datasets.

**Run on all 4 datasets** (use the winning decay from TandemFoil for the other 3):

### TandemFoil — run both decay=0.8 and decay=0.9
```bash
cd target/icml2026
# decay=0.8
python train.py \
  --dataset tandemfoil \
  --optimizer lion --lr 1e-4 \
  --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame \
  --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 \
  --wandb_group llrd-cross-dataset

# decay=0.9
python train.py \
  --dataset tandemfoil \
  --optimizer lion --lr 1e-4 \
  --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame \
  --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 \
  --wandb_group llrd-cross-dataset
```

### AirfRANS (use winning decay)
```bash
cd target/icml2026
python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 10 --grad-clip 1.0 --no-use-ema \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb_group llrd-cross-dataset
```

### DrivAerML (use winning decay)
```bash
cd target/icml2026
SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --no-use-ema \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group llrd-cross-dataset
```

### TandemFoil Paper (use winning decay)
```bash
cd target/icml2026
python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion --lr 1e-4 \
  --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame \
  --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 \
  --wandb_group llrd-cross-dataset
```

## Reporting

In your PR comment, report:
- W&B run IDs for all runs (including both TandemFoil decay variants)
- `val_primary/*` and `test_primary/*` metrics for each dataset (test from best val checkpoint)
- Which decay value won on TandemFoil and why you chose it for the other datasets
- Effective LR per layer at the start of training (log to W&B or print in comment)

## Notes

- For Lion optimizer (TandemFoil/TandemFoil Paper), Lion accepts param groups with different LRs just like AdamW.
- For a 3-layer model (TandemFoil), decay=0.8 gives LRs: `[0.8^3, 0.8^2, 0.8, 1.0] * base_lr = [0.51e-4, 0.64e-4, 0.8e-4, 1.0e-4]`.
- tandemfoil_paper has no established baseline — this run establishes it.
- If the model uses a non-standard layer structure (e.g. MoE-FFN blocks), apply LLRD to the entire block as a unit.

### CODE REVIEW

**Comment by morganmcg1** [comment]
Closing to narrow the queue for the final ICML benchmark sprint. Current priority is benchmark-specific optimization and benchmark-faithful reporting for DrivAerML, TandemFoil paper benchmark, AirfRANS full Surf/Vol, and the Tandem parity anchor. If this idea still matters later, we can reopen it in a benchmark-specific form.
